core/src/main/java/org/infinispan/marshall/core/internal/BytesObjectOutput.java

```diff
@@ -0,0 +1,131 @@
+package org.infinispan.marshall.core.internal;
+
+import org.infinispan.commons.marshall.Externalizer;
+
+import java.io.IOException;
+import java.io.ObjectOutput;
+
+final class BytesObjectOutput implements ObjectOutput, PositionalBuffer.Output {
+
+   final InternalMarshaller internal;
+
+   byte bytes[];
+   int pos;
+
+   public BytesObjectOutput(int size, InternalMarshaller internal) {
+      this.bytes = new byte[size];
+      this.internal = internal;
+   }
+
+   @Override
+   public void writeObject(Object obj) throws IOException {
```

Comment:
I think it would be great if we took this opportunity to add a `void writeSharedObject(Object)` method that did the work of `InstanceReusingAdvancedExternalizers`, but without using thread-locals. Or maybe `bool writeObjectReference(Object)`, so an externalizer could start with `if (writeObjectReference(obj)) return;`?